source/scpd/tests/LayoutTest.d

```diff
@@ -0,0 +1,80 @@
+/*******************************************************************************
+
+    Contains runtime size & ABI object layout checks and
+    function argument passing in the D glue layer works properly.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module scpd.tests.LayoutTest;
+
+import scpd.tests.GlueTypes;
+import scpd.types.Stellar_SCP;
+import scpd.types.Stellar_types;
+
+import std.digest.sha;
+import std.random;
+import std.stdio;
+import std.string;
+import std.traits;
+
+/// Contains the size and offset of a field within a struct
+extern(C++) struct FieldInfo
+{
+    long size;
+    long offset;
+    //const(char)* mangleof;  // todo: could this work?
+}
+
+static foreach (Type; GlueTypes)
+{
+    extern(C++) FieldInfo cppFieldInfo (ref Type, const(char)*);
+}
+
+/// Special-case as we implement our own Hash type that is binary-compatible with SCP's Hash
+/// Note: cannot pass with proper type due to mangling issues:
+///  "isHashTypeCompatible(BitBlob<256ul>&)", referenced from:
+///      __D4scpd5tests10LayoutTest17__unittest_L43_C1FZv in agora-test-application.o
```

Comment:
One question I have is: "how does mangling behave w.r.t `alias`. Aliases are transparent, so we should probably use `struct Name { RealType data; alias data this; }` however it has its problems (namely there's a Name -> RealType path but not the other way around).

Reply:
Yeah I think we might need to do this instead of aliases in the binding. Care to try ?

Reply:
Hmm, are there any benefits to this? Because implicit conversions will still happen. Genuinely curious as I rarely use subtyping.

Reply:
Well if you declare a `void fun(ref Hash)` function on the D side and a `void fun(Hash&)` on the C++ side and it doesn't link I'd say it's a pretty annoying bug.

Reply:
ok so you want me to try declaring Hash as:

```D
struct Hash
{
    BitBlob!(...) _internal;
    alias _internal this;
}
```

And see how far we can take this?